Add `isRequired` flag option

sbencoding opened the pull request:
This PR aims to add the ability to specify an `isRequired` option for flags, based on code and feedback from #110 and #51.
#### What I did
* Implement the `isRequired` option. Can be a `boolean` or a `function`.
* Update readme and type definition with the `isRequired` option.
* Add tests and new fixture files for the new feature
#### Questions/suggestions
I think it's enough to pass the `flags` and the `input` variables to the `isRequired` function. That should give the consuming application enough context  to decide the return value, at least from the package's point of view IMO.

In the original issue it was suggested that this should be entirely handled by the consuming application. I didn't know if I should, but it's possible to add an additional flag that changes the behavior of `isRequired`, so that instead of exiting and printing the missing options, it would include the missing flags in the result returned by the API.

I didn't add any additional tests to `index.test-d.ts` because I wasn't sure what to test for. `isRequired` either closes the app if some value is missing or leaves the value untouched if it's specified, so I don't know if the feature needs any type related testing.

Let me know what you think and if I should change anything.

Fixes #51


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#51: Support `required` flags](https://issuehunt.io/repos/25121561/issues/51)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

sindresorhus commented on index.js:
```diff
@@ -118,6 +128,30 @@ const meow = (helpText, options) => {
+
+		process.exit(2);
+	}
+
```
This main function is getting huge. Would be great if you could modularize the logic you're adding into separate methods.

sindresorhus commented on readme.md:
```diff
@@ -97,6 +97,7 @@ The key is the flag name and the value is an object with any of:
 - `type`: Type of value. (Possible values: `string` `boolean` `number`)
 - `alias`: Usually used to define a short flag alias.
 - `default`: Default value when the flag is not specified.
+- `isRequired`: Boolean or Function that specified is this flag is required.
```
This needs to describe the parameters the function receives.

sindresorhus commented on readme.md:
```diff
@@ -106,6 +107,9 @@ flags: {
 		alias: 'u',
 		default: 'rainbow'
+		isRequired: (flags, input) => {
+			if (flags.otherFlag) return true;
```
Use braces

sindresorhus commented on index.js:
```diff
@@ -118,6 +128,30 @@ const meow = (helpText, options) => {
+	if (missing.length > 0) {
+		console.log(`Missing required option${missing.length > 1 ? 's' : ''}`);
+		for (const flag of missing) {
+			console.log(`\t--${flag.key}${flag.alias ? `, -${flag.alias}` : ''}`);
```
Shouldn't this be `console.error`?

sindresorhus commented on index.js:
```diff
@@ -118,6 +128,30 @@ const meow = (helpText, options) => {
+
+	// Print error message for missing flags
+	if (missing.length > 0) {
+		console.log(`Missing required option${missing.length > 1 ? 's' : ''}`);
```
I think we should be consistent with the word "flag" instead of "option".

sindresorhus commented on index.js:
```diff
@@ -118,6 +128,30 @@ const meow = (helpText, options) => {
 	}
 
+	// Get a list of missing flags, that are required
+	const missing = [];
```
"missing" what? Needs a more descriptive name.

sindresorhus commented on fixture-required.js:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env node
```
There's a lot of fixtures now. Can you move them into a `test/fixtures` directory?

sindresorhus commented on test.js:
```diff
@@ -80,6 +80,72 @@ test('spawn cli and test input flag', async t => {
 	t.is(stdout, 'bar');
 });
 
+test('spawn cli and test required flag with missing args', async t => {
```
Can you move these new tests into a new test file? And also move all tests into a `test` directory?

sbencoding commented:
I've implemented the requested changes and merged with the latest master as well, let me know if there's anything else to change.

sindresorhus commented on index.d.ts:
```diff
@@ -2,11 +2,13 @@ import {PackageJson} from 'type-fest';
 
 declare namespace meow {
 	type FlagType = 'string' | 'boolean' | 'number';
+	type IsRequiredPredicate = (flags: AnyFlags, input: string[]) => boolean;
```
```suggestion
	type IsRequiredPredicate = (flags: AnyFlags, input: readonly string[]) => boolean;
```

sindresorhus commented on readme.md:
```diff
@@ -97,6 +97,11 @@ The key is the flag name and the value is an object with any of:
+- `isRequired`: Boolean or Function that specifies if this flag is required.
+	Two arguments are passed to the function.
+	The first arguments is the **flags** object, it contains the flags converted to camelCase excluding aliases.
+	The second arugment is the **input** string array, it contains the non-flag arguments.
```
typo

sindresorhus commented on readme.md:
```diff
@@ -97,6 +97,11 @@ The key is the flag name and the value is an object with any of:
+	Two arguments are passed to the function.
+	The first arguments is the **flags** object, it contains the flags converted to camelCase excluding aliases.
+	The second arugment is the **input** string array, it contains the non-flag arguments.
+	The function should return a Boolean, true if the flag is requried, otherwise false.
```
typo

sindresorhus commented on readme.md:
```diff
@@ -106,6 +111,11 @@ flags: {
+		isRequired: (flags, input) => {
+			if (flags.otherFlag) {
+				return true;
+			}
```
This does not follow the docs as it doesn't return false for the negative case.

sindresorhus replied:
Should `isRequired` throw if the return value is not a boolean?

sbencoding replied:
I've decided to implement a check for this that throws if the value returned is not a boolean.
I think it's a good idea, because just as I missed the negative condition in the example above, somebody might forget to return the proper values, and that might lead to some unexpected results.

sindresorhus commented on index.d.ts:
```diff
@@ -24,6 +26,11 @@ declare namespace meow {
+			Two arguments are passed to the function.
+			The first arguments is the flags object, it contains the flags converted to camelCase excluding aliases.
+			The second arugment is the input string array, it contains the non-flag arguments.
+			The function should return a Boolean, true if the flag is requried, otherwise false.
```
The function docs should be at https://github.com/sindresorhus/meow/pull/141/files#diff-b52768974e6bc0faccb7d4b75b162c99R5 and you don't have to mention the types. It's already shown by the actual types. Mention what it does and what problem it solves.

sbencoding replied:
Yeah, I was just trying to make sure that the `index.d.ts` and the `readme.md` file reflect each other as much as possible, that's why I left the type information there, because although typescript has a better way of documenting type information, in the readme file this is the only way.
Anyways I left the documentation with the type information in the readme file and updated the typescript file with a better documentation

sindresorhus commented on test/test-is-required-flag.js:
```diff
@@ -0,0 +1,72 @@
+import test from 'ava';
```
The filename can just be `is-required-flag.js`

sindresorhus commented on index.js:
```diff
@@ -14,6 +14,37 @@ const normalizePackageData = require('normalize-package-data');
+
+const getMissingRequiredFlags = (flags, receivedFlags, input) => {
+	const missingRequiredFlags = [];
+	if (typeof flags !== 'undefined') {
```
Use an early-return here to reduce the nesting.

sbencoding commented:
I've implemented the requested changes and merged with the current upstream master branch that contains the `isMultiple` flag.
Because of this merge I've added a few more test to make sure `isRequired` plays nicely with it.

sindresorhus commented:
Looks good. Thank :)
